app/code/Magento/Cron/Model/ResourceModel/Schedule/ExpressionValidator.php

```diff
@@ -0,0 +1,213 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Cron\Model\ResourceModel\Schedule;
+
+use Magento\Cron\Model\Schedule;
+use Magento\Framework\Exception\CronException;
+
+/**
+ * Schedule cron expression validator
+ *
+ * @api
+ */
+class ExpressionValidator
+{
+    const CRON_EXPRESSION_MIN_PARTS_NUMBER = 5;
+    const CRON_EXPRESSION_MAX_PARTS_NUMBER = 6;
+
+    /**
+     * @var Schedule
+     */
+    private $schedule;
+
+    /**
+     * @var string
+     */
+    private $lastErrorMsg;
+
+    /**
+     * @var array
+     */
+    private $fromToModulus;
+
+    /**
+     * @param Schedule $schedule
+     */
+    public function setSchedule($schedule)
+    {
+        $this->schedule = $schedule;
+    }
+
+    /**
+     * @return string
+     */
+    public function getLastErrorMsg()
+    {
+        return $this->lastErrorMsg;
+    }
+
+    /**
+     * Checks the schedule's cron expression validity
+     *
+     * @return bool
+     */
+    public function checkCronExpr()
+    {
+        $expr = $this->schedule->getCronExprArr();
+
+        if (!$expr) {
+            return false;
+        }
+
+        if (sizeof($expr) < self::CRON_EXPRESSION_MIN_PARTS_NUMBER
+            || sizeof($expr) > self::CRON_EXPRESSION_MAX_PARTS_NUMBER
+        ) {
+            return false;
+        }
+
+        $check = $this->checkCronExpressionPart($expr[0])
+            && $this->checkCronExpressionPart($expr[1])
+            && $this->checkCronExpressionPart($expr[2])
+            && $this->checkCronExpressionPart($expr[3])
+            && $this->checkCronExpressionPart($expr[4]);
+
+        return $check;
+    }
+
+    /**
+     * @param string $expr
+     * @return bool
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    public function checkCronExpressionPart($expr)
+    {
+        $this->lastErrorMsg = '';
+        $this->fromToModulus = [];
+
+        // handle ALL match
+        if ($this->isCronExpressionPartAsterisk($expr)) {
+            return true;
+        }
+
+        // handle multiple options
+        if ($this->isCronExpressionPartList($expr)) {
+            foreach (explode(',', $expr) as $e) {
+                if ($this->checkCronExpressionPart($e)) {
+                    return true;
```

Comment:
Error in logic. The loop should be interrupted if some part is invalid. And if all items from the list are valid then list is valid.